src/EditorFeatures/TestUtilities/TestExportJoinableTaskContext.cs

```diff
@@ -1,24 +1,19 @@
 ﻿using System;
-using System.Collections.Generic;
 using System.ComponentModel.Composition;
-using System.Linq;
-using System.Text;
-using System.Threading;
-using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
 using Microsoft.VisualStudio.Threading;
 
 namespace Microsoft.CodeAnalysis.Editor.UnitTests
 {
-    // In 15.3 the editor took a dependency on JoinableTaskContext.
-    // JTC appears in the VS MEF composition but does not itself
-    // contain an ExportAttribute. The Editor's own unit tests
-    // export it from a field and we need to do the same in order
-    // to be able to compose in any part of the Editor.
+    // Starting with 15.3 the editor took a dependency on JoinableTaskContext
+    // in Text.Logic and Intellisense layers as an editor host provided service.
+    [PartCreationPolicy(CreationPolicy.NonShared)] // JTC is "main thread" affinitized so should not be shared
```

Comment:
The changes to this file seem somewhat fishy to me:

1. You're no longer using the CurrentForegroundThreadData which means this type doesn't need to inherit from ForegroundThreadAffinitizedObject anymore.
2. Because of the previous point, there's no real reason to create an instance of this type at all...which seems that the creation policy doesn't matter.
3. I don't see why this would be ThreadStatic, as we really shouldn't have tests that care about this then other than the ones that use our WpfFact to run on a single thread.
4. The part creation policy should only matter in a single composition, which wouldn't have an impact on if it's being used across threads.

Maybe we have to just chat in person on this one, but what's going on here? This seems very magic and there's a lot of context missing.

Reply:
let me look at it again...

Reply:
it does look fishy, I wish I knew a better way.
Essentially what I found:
1. Without non-shared part creation policy most most tests (specifically I believe non WpfFact ones) creating text view fail because new QuickInfo text view creation listener throws for not being on the main thread. I think that's because JTC instances are cached by MinimalTestExportProvider between compositions and so being reused on other threads.
2. Without [ThreadStatic] tests deadlock, presumably because we end up creating more than one JTC for same thread.

This needs further investigation, I just don't have enough context in the text infrastructure. Maybe it would be better not to export JTC here, but explicitly add new instance to each composition?
I suggest creating an issue to track it, but not block this PR. I don't mind spending more time on this.